HFE.md

```diff
@@ -0,0 +1,643 @@
+
+## Motivation
+
+Introducing field-level expiration for data structures like hashes and sets in Valkey would significantly enhance its flexibility and applicability in modern application architectures. Many use cases—such as session management, per-user rate limits, temporary access rights, feature store toggles, and short-lived membership in groups—require storing multiple items within a single Valkey key, where each item has its own independent expiration time. Currently, Valkey supports expiration only at the key level, which forces developers to either fragment data across multiple keys or implement complex application-side logic to simulate item-level TTLs. This increases operational overhead, complicates data access patterns, and undermines the simplicity Valkey is known for. Native support for field or member expiration within hashes and sets would streamline data modeling, reduce maintenance burdens, and enable more efficient memory usage and access control in real-time systems.
+
+Sub-Items expiration was introduced in several Redis-compatible forks including [Keydb](https://docs.keydb.dev/blog/2019/10/21/blog-post/), [Dragonfly](https://www.dragonflydb.io/docs/command-reference/hashes/hsetex), [Redis](https://redis.io/blog/hash-field-expiration-architecture-and-benchmarks/) and [TairHash module ](https://github.com/tair-opensource/TairHash).
+In order to unblock workloads which are utilizing this logic we would like valkey to offer a similar ability.
+
+## Design considerations 
+
+### Requirements
+
+* Support Redis compatible API to set, get and manipulate Hash field TTL.
+* Support both Lazy and active expiry mechanisms for Hash field expirations.
+* Support Replication of elements TTL as well as expired element replication.
+* Extended support for the same functionality with Sets and Sorted sets.
+
+### Tenets
+
+* **memory efficiency** - At the highest priority we target a minimal metadata overhead used in order to store and manage items TTL. While the optimal overhead to maintain item TTL is 8 bytes (could be less if we allow keeping offsets from the existing epoch diff time), we understand that maintaining active expiry logic will require use of more bytes for metadata. We will make our top priority effort to minimize this overhead.
+* **Latency** - After memory efficiency considerations we will require a solution which provides low latency for hash operations. Hash objects are expected to operate in O(1) for all single access operations (get, set, delete etc...) and we will not break this promise even for items with expiry.
```

Comment:
"provides low latency" is vague. Do you mean 'doesn’t regress existing latency by a noticeable margin'?

Reply:
IMO I was clear about asymptotic guarantees.